File: tests/aidl/default/TestAidlMsgQ.cpp
/*
 * Copyright (C) 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "TestAidlMsgQ.h"

namespace aidl {
namespace android {
namespace fmq {
namespace test {

template <typename T>
ndk::ScopedAStatus configureFmqSyncReadWriteGeneric(
        const MQDescriptor<T, SynchronizedReadWrite>& desc,
        std::unique_ptr<TestAidlMsgQ::MessageQueueSync<T>>& fmq) {
    fmq.reset(new (std::nothrow) TestAidlMsgQ::MessageQueueSync(desc));
    if ((fmq == nullptr) || (fmq->isValid() == false)) {
        return ndk::ScopedAStatus::fromExceptionCode(EX_ILLEGAL_ARGUMENT);
    }

    /* Initialize the EventFlag word with bit FMQ_NOT_FULL. */
    auto evFlagWordPtr = fmq->getEventFlagWord();
    if (evFlagWordPtr != nullptr) {
        std::atomic_init(evFlagWordPtr, static_cast<uint32_t>(EventFlagBits::FMQ_NOT_FULL));
    }
    return ndk::ScopedAStatus::ok();
}

template <typename T>
bool getFmqUnsyncWriteGeneric(bool configureFmq, bool userFd,
                              MQDescriptor<T, UnsynchronizedWrite>* mqDesc,
                              std::unique_ptr<TestAidlMsgQ::MessageQueueUnsync<T>>& fmqUnsync) {
    if (configureFmq) {
        static constexpr size_t kNumElementsInQueue = 1024;
        static constexpr size_t kElementSizeBytes = sizeof(int32_t);
        ::android::base::unique_fd ringbufferFd;
        if (userFd) {
            ringbufferFd.reset(
                    ::ashmem_create_region("UnsyncWrite", kNumElementsInQueue * kElementSizeBytes));
        }
        fmqUnsync.reset(new (std::nothrow) TestAidlMsgQ::MessageQueueUnsync<T>(
                kNumElementsInQueue, false, std::move(ringbufferFd),
                kNumElementsInQueue * kElementSizeBytes));
    }

    if ((fmqUnsync == nullptr) || (fmqUnsync->isValid() == false) || (mqDesc == nullptr)) {
        return false;
    } else {
        *mqDesc = fmqUnsync->dupeDesc();
        // set write-protection so readers can't mmap and write
        int res = ashmem_set_prot_region(mqDesc->handle.fds[0].get(), PROT_READ);
        if (res == -1) {
            ALOGE("Failed to set write protection: %s", strerror(errno));
            return false;
        } else {
            return true;
        }
    }
}

// Methods from ::aidl::android::fmq::test::ITestAidlMsgQ follow.
ndk::ScopedAStatus TestAidlMsgQ::configureFmqSyncReadWrite(
        const MQDescriptor<int32_t, SynchronizedReadWrite>& mqDesc, bool* _aidl_return) {
    auto status = configureFmqSyncReadWriteGeneric(mqDesc, mFmqSynchronized);
    *_aidl_return = true;
    return status;
}

ndk::ScopedAStatus TestAidlMsgQ::getFmqUnsyncWrite(
        bool configureFmq, bool userFd, MQDescriptor<int32_t, UnsynchronizedWrite>* mqDesc,
        bool* _aidl_return) {
    *_aidl_return = getFmqUnsyncWriteGeneric(configureFmq, userFd, mqDesc, mFmqUnsynchronized);
    return ndk::ScopedAStatus::ok();
}

ndk::ScopedAStatus TestAidlMsgQ::configureFmqAidlTypesSyncReadWrite(
        const std::optional<MQDescriptor<ITestAidlMsgQ::StructPayload, SynchronizedReadWrite>>&
                in_mqDescStruct,
        const std::optional<MQDescriptor<ITestAidlMsgQ::UnionPayload, SynchronizedReadWrite>>&
                in_mqDescUnion,
        const std::optional<MQDescriptor<ITestAidlMsgQ::EnumPayload, SynchronizedReadWrite>>&
                in_mqDescEnum,
        bool* _aidl_return) {
    /* enforce that exactly one argument is non-empty. */
    int n_nonempty =
            in_mqDescStruct.has_value() + in_mqDescUnion.has_value() + in_mqDescEnum.has_value();
    if (n_nonempty != 1) {
        return ndk::ScopedAStatus::fromExceptionCode(EX_ILLEGAL_ARGUMENT);
    }

    ndk::ScopedAStatus status;
    if (in_mqDescStruct.has_value()) {
        status = configureFmqSyncReadWriteGeneric(*in_mqDescStruct, mFmqSynchronizedStruct);
    }
    if (in_mqDescUnion.has_value()) {
        status = configureFmqSyncReadWriteGeneric(*in_mqDescUnion, mFmqSynchronizedUnion);
    }
    if (in_mqDescEnum.has_value()) {
        status = configureFmqSyncReadWriteGeneric(*in_mqDescEnum, mFmqSynchronizedEnum);
    }

    if (status.isOk()) {
        *_aidl_return = true;
    }
    return status;
}

// TODO: verify that this is treating AIDL nullability vs out-ptrs correctly; it may not be
ndk::ScopedAStatus TestAidlMsgQ::getFmqAidlTypesUnsyncWrite(
        bool in_configureFmq, bool in_userFd,
        std::optional<MQDescriptor<ITestAidlMsgQ::StructPayload, UnsynchronizedWrite>>*
                out_mqDescStruct,
        std::optional<MQDescriptor<ITestAidlMsgQ::UnionPayload, UnsynchronizedWrite>>*
                out_mqDescUnion,
        std::optional<MQDescriptor<ITestAidlMsgQ::EnumPayload, UnsynchronizedWrite>>*
                out_mqDescEnum,
        bool* _aidl_return) {
    /* enforce that exactly one argument is non-empty. */
    int n_nonempty = out_mqDescStruct->has_value() + out_mqDescUnion->has_value() +
                     out_mqDescEnum->has_value();
    if (n_nonempty != 1) {
        return ndk::ScopedAStatus::fromExceptionCode(EX_ILLEGAL_ARGUMENT);
    }

    if (out_mqDescStruct->has_value()) {
        *_aidl_return = getFmqUnsyncWriteGeneric(
                in_configureFmq, in_userFd, &out_mqDescStruct->value(), mFmqUnsynchronizedStruct);
    }
    if (out_mqDescUnion->has_value()) {
        *_aidl_return = getFmqUnsyncWriteGeneric(
                in_configureFmq, in_userFd, &out_mqDescUnion->value(), mFmqUnsynchronizedUnion);
    }
    if (out_mqDescEnum->has_value()) {
        *_aidl_return = getFmqUnsyncWriteGeneric(in_configureFmq, in_userFd,
                                                 &out_mqDescEnum->value(), mFmqUnsynchronizedEnum);
    }
    return ndk::ScopedAStatus::ok();
}

ndk::ScopedAStatus TestAidlMsgQ::requestBlockingRead(int32_t count) {
    bool result;
    if (mFmqSynchronized != nullptr) {
        std::vector<int32_t> data(count);
        result = mFmqSynchronized->readBlocking(
                &data[0], count, static_cast<uint32_t>(EventFlagBits::FMQ_NOT_FULL),
                static_cast<uint32_t>(EventFlagBits::FMQ_NOT_EMPTY), 5000000000 /* timeOutNanos */);
    }
    if (mFmqSynchronizedStruct != nullptr) {
        std::vector<StructPayload> data(count);
        result = mFmqSynchronizedStruct->readBlocking(
                &data[0], count, static_cast<uint32_t>(EventFlagBits::FMQ_NOT_FULL),
                static_cast<uint32_t>(EventFlagBits::FMQ_NOT_EMPTY), 5000000000 /* timeOutNanos */);
    }
    if (mFmqSynchronizedEnum != nullptr) {
        std::vector<EnumPayload> data(count);
        result = mFmqSynchronizedEnum->readBlocking(
                &data[0], count, static_cast<uint32_t>(EventFlagBits::FMQ_NOT_FULL),
                static_cast<uint32_t>(EventFlagBits::FMQ_NOT_EMPTY), 5000000000 /* timeOutNanos */);
    }
    if (mFmqSynchronizedUnion != nullptr) {
        std::vector<UnionPayload> data(count);
        result = mFmqSynchronizedUnion->readBlocking(
                &data[0], count, static_cast<uint32_t>(EventFlagBits::FMQ_NOT_FULL),
                static_cast<uint32_t>(EventFlagBits::FMQ_NOT_EMPTY), 5000000000 /* timeOutNanos */);
    }

    if (result == false) {
        ALOGE("Blocking read fails");
    }
    return ndk::ScopedAStatus::ok();
}

ndk::ScopedAStatus TestAidlMsgQ::requestBlockingReadDefaultEventFlagBits(int32_t count) {
    std::vector<int32_t> data(count);
    bool result;
    if (mFmqSynchronized != nullptr) {
        std::vector<int32_t> data(count);
        result = mFmqSynchronized->readBlocking(&data[0], count);
    }
    if (mFmqSynchronizedStruct != nullptr) {
        std::vector<StructPayload> data(count);
        result = mFmqSynchronizedStruct->readBlocking(&data[0], count);
    }
    if (mFmqSynchronizedEnum != nullptr) {
        std::vector<EnumPayload> data(count);
        result = mFmqSynchronizedEnum->readBlocking(&data[0], count);
    }
    if (mFmqSynchronizedUnion != nullptr) {
        std::vector<UnionPayload> data(count);
        result = mFmqSynchronizedUnion->readBlocking(&data[0], count);
    }

    if (result == false) {
        ALOGE("Blocking read fails");
    }

    return ndk::ScopedAStatus::ok();
}

ndk::ScopedAStatus TestAidlMsgQ::requestBlockingReadRepeat(int32_t count, int32_t numIter) {
    for (int i = 0; i < numIter; i++) {
        requestBlockingRead(count);
    }

    return ndk::ScopedAStatus::ok();
}

ndk::ScopedAStatus TestAidlMsgQ::requestReadFmqSync(int32_t count, bool* _aidl_return) {
    // TODO: port verifyData to other types
    bool result;
    if (mFmqSynchronized != nullptr) {
        std::vector<int32_t> data(count);
        result = mFmqSynchronized->read(&data[0], count) && verifyData(&data[0], count);
    }
    if (mFmqSynchronizedStruct != nullptr) {
        std::vector<StructPayload> data(count);
        result = mFmqSynchronizedStruct->read(&data[0], count);
    }
    if (mFmqSynchronizedEnum != nullptr) {
        std::vector<EnumPayload> data(count);
        result = mFmqSynchronizedEnum->read(&data[0], count);
    }
    if (mFmqSynchronizedUnion != nullptr) {
        std::vector<UnionPayload> data(count);
        result = mFmqSynchronizedUnion->read(&data[0], count);
    }

    *_aidl_return = result;
    return ndk::ScopedAStatus::ok();
}

ndk::ScopedAStatus TestAidlMsgQ::requestReadFmqUnsync(int32_t count, bool* _aidl_return) {
    // TODO: port verifyData to other types
    bool result;
    if (mFmqUnsynchronized != nullptr) {
        std::vector<int32_t> data(count);
        result = mFmqUnsynchronized->read(&data[0], count) && verifyData(&data[0], count);
    }
    if (mFmqUnsynchronizedStruct != nullptr) {
        std::vector<StructPayload> data(count);
        result = mFmqUnsynchronizedStruct->read(&data[0], count);
    }
    if (mFmqUnsynchronizedEnum != nullptr) {
        std::vector<EnumPayload> data(count);
        result = mFmqUnsynchronizedEnum->read(&data[0], count);
    }
    if (mFmqUnsynchronizedUnion != nullptr) {
        std::vector<UnionPayload> data(count);
        result = mFmqUnsynchronizedUnion->read(&data[0], count);
    }

    *_aidl_return = result;
    return ndk::ScopedAStatus::ok();
}

ndk::ScopedAStatus TestAidlMsgQ::requestWriteFmqSync(int32_t count, bool* _aidl_return) {
    // TODO: port to other types
    std::vector<int32_t> data(count);
    for (int i = 0; i < count; i++) {
        data[i] = i;
    }
    bool result = mFmqSynchronized->write(&data[0], count);
    *_aidl_return = result;
    return ndk::ScopedAStatus::ok();
}

ndk::ScopedAStatus TestAidlMsgQ::requestWriteFmqUnsync(int32_t count, bool* _aidl_return) {
    // TODO: port to other types
    std::vector<int32_t> data(count);
    for (int i = 0; i < count; i++) {
        data[i] = i;
    }
    if (!mFmqUnsynchronized) {
        ALOGE("Unsynchronized queue is not configured.");
        *_aidl_return = false;
        return ndk::ScopedAStatus::ok();
    }
    bool result = mFmqUnsynchronized->write(&data[0], count);
    *_aidl_return = result;
    return ndk::ScopedAStatus::ok();
}

}  // namespace test
}  // namespace fmq
}  // namespace android
}  // namespace aidl

File: tests/aidl/default/TestAidlMsgQ.rs
//! This module implements the ITestAidlMsgQ AIDL interface

/*
* Copyright (C) 2024 The Android Open Source Project
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*      http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

use std::{fmt::Debug, sync::Mutex};

use android_fmq_test::aidl::android::fmq::test::{
    EventFlagBits::EventFlagBits,
    ITestAidlMsgQ::{
        EnumPayload::EnumPayload, ITestAidlMsgQ, StructPayload::StructPayload,
        UnionPayload::UnionPayload,
    },
};
use android_fmq_test::binder::{self, Interface, Result as BinderResult};

/// Struct implementing the ITestAidlMsgQ AIDL interface
#[derive(Default)]
pub struct MsgQTestService {
    queue_sync_struct: Mutex<Option<fmq::MessageQueue<StructPayload>>>,
    queue_sync_union: Mutex<Option<fmq::MessageQueue<UnionPayload>>>,
    queue_sync_enum: Mutex<Option<fmq::MessageQueue<EnumPayload>>>,
    queue_sync: Mutex<Option<fmq::MessageQueue<i32>>>,
}

impl Interface for MsgQTestService {}

use android_hardware_common_fmq::aidl::android::hardware::common::fmq::{
    MQDescriptor::MQDescriptor, SynchronizedReadWrite::SynchronizedReadWrite,
    UnsynchronizedWrite::UnsynchronizedWrite,
};

use std::sync::atomic::Ordering;

/// Create a `MessageQueue` from the given descriptor, configuring an event flag
/// word and initializing it to `EventFlagBits::FMQ_NOT_FULL`.
fn mq_with_ev_flag_word<T>(desc: &MQDescriptor<T, SynchronizedReadWrite>) -> fmq::MessageQueue<T>
where
    T: zerocopy::TryFromBytes,
{
    let mq = fmq::MessageQueue::from_desc(desc, true);
    /* Set the EventFlag word with bit FMQ_NOT_FULL. */
    if let Some(event_word) = mq.event_flag_word() {
        event_word.store(EventFlagBits::FMQ_NOT_FULL.0 as u32, Ordering::Relaxed);
    }
    mq
}

impl ITestAidlMsgQ for MsgQTestService {
    /**
     * This method requests the service to set up a synchronous read/write
     * wait-free FMQ using the input descriptor with the client as reader.
     *
     * @param mqDesc This structure describes the FMQ that was set up by the
     * client. Server uses this descriptor to set up a FMQ object at its end.
     *
     * @return True if the setup is successful.
     */
    fn configureFmqSyncReadWrite(
        &self,
        mq_desc: &MQDescriptor<i32, SynchronizedReadWrite>,
    ) -> BinderResult<bool> {
        *self.queue_sync_struct.lock().unwrap() = None;
        *self.queue_sync_union.lock().unwrap() = None;
        *self.queue_sync_enum.lock().unwrap() = None;
        *self.queue_sync.lock().unwrap() = Some(mq_with_ev_flag_word(mq_desc));

        Ok(true)
    }

    /**
     * This method requests the service to set up a synchronous read/write
     * wait-free FMQ using the given input descriptor with client as reader.
     *
     * Only exactly one of the arguments should be non-NULL. The server will use
     * the non-NULL descriptor to set up a FMQ object at its end, which can then
     * be directed to read or write with the various request* methods.
     *
     * @return True if the setup is successful.
     */
    fn configureFmqAidlTypesSyncReadWrite(
        &self,
        mq_desc_struct: Option<&MQDescriptor<StructPayload, SynchronizedReadWrite>>,
        mq_desc_union: Option<&MQDescriptor<UnionPayload, SynchronizedReadWrite>>,
        mq_desc_enum: Option<&MQDescriptor<EnumPayload, SynchronizedReadWrite>>,
    ) -> BinderResult<bool> {
        assert!(
            [mq_desc_struct.is_some(), mq_desc_union.is_some(), mq_desc_enum.is_some(),]
                .into_iter()
                .filter(|b| *b)
                .count()
                == 1,
            "exactly one descriptor must be non-NULL"
        );
        *self.queue_sync_struct.lock().unwrap() = mq_desc_struct.map(mq_with_ev_flag_word);
        *self.queue_sync_union.lock().unwrap() = mq_desc_union.map(mq_with_ev_flag_word);
        *self.queue_sync_enum.lock().unwrap() = mq_desc_enum.map(mq_with_ev_flag_word);
        *self.queue_sync.lock().unwrap() = None;
        Ok(true)
    }
    fn getFmqAidlTypesUnsyncWrite(
        &self,
        _: bool,
        _: bool,
        _: &mut Option<MQDescriptor<StructPayload, UnsynchronizedWrite>>,
        _: &mut Option<MQDescriptor<UnionPayload, UnsynchronizedWrite>>,
        _: &mut Option<MQDescriptor<EnumPayload, UnsynchronizedWrite>>,
    ) -> BinderResult<bool> {
        // The Rust interface to FMQ does not support `UnsynchronizedWrite`.
        Ok(false)
    }

    /**
     * This method requests the service to read from the synchronized read/write
     * FMQ.
     *
     * @param count Number to messages to read.
     *
     * @return True if the read operation was successful.
     */
    fn requestReadFmqSync(&self, count: i32) -> BinderResult<bool> {
        fn read_sync<T: zerocopy::TryFromBytes + Debug>(
            mutex_mq: &Mutex<Option<fmq::MessageQueue<T>>>,
            count: i32,
        ) -> Option<bool> {
            let mut queue_guard = mutex_mq.lock().unwrap();
            let mq = queue_guard.as_mut()?;
            let rc = mq.read_many(count.try_into().unwrap());
            match rc {
                Some(mut rc) => {
                    for _ in 0..count {
                        rc.try_read().unwrap().unwrap();
                    }
                    Some(true)
                }
                None => {
                    eprintln!("failed to read_many({count})");
                    Some(false)
                }
            }
        }

        if let Some(ret) = read_sync(&self.queue_sync_struct, count) {
            return Ok(ret);
        }
        if let Some(ret) = read_sync(&self.queue_sync_union, count) {
            return Ok(ret);
        }
        if let Some(ret) = read_sync(&self.queue_sync_enum, count) {
            return Ok(ret);
        }
        if let Some(ret) = read_sync(&self.queue_sync, count) {
            return Ok(ret);
        }

        Err(binder::Status::new_service_specific_error_str(107, Some("no fmq set up")))
    }

    /**
     * This method requests the service to write into the synchronized read/write
     * flavor of the FMQ.
     *
     * @param count Number to messages to write.
     *
     * @return True if the write operation was successful.
     */
    fn requestWriteFmqSync(&self, count: i32) -> BinderResult<bool> {
        fn write_sync<T: binder::WriteTo + zerocopy::Immutable + Debug, F: Fn(i32) -> T>(
            mutex_mq: &Mutex<Option<fmq::MessageQueue<T>>>,
            count: i32,
            gen_val: F,
        ) -> Option<bool> {
            let mut queue_guard = mutex_mq.lock().unwrap();
            let mq = queue_guard.as_mut()?;
            let wc = mq.write_many(count.try_into().unwrap());
            match wc {
                Some(mut wc) => {
                    for i in 0..count {
                        wc.write(gen_val(i)).unwrap();
                    }
                    drop(wc);
                    Some(true)
                }
                None => {
                    eprintln!("failed to write_many({count})");
                    Some(false)
                }
            }
        }

        if let Some(ret) = write_sync(&self.queue_sync_struct, count, |_: i32| Default::default()) {
            return Ok(ret);
        }
        if let Some(ret) = write_sync(&self.queue_sync_union, count, |_: i32| Default::default()) {
            return Ok(ret);
        }
        if let Some(ret) = write_sync(&self.queue_sync_enum, count, |_: i32| Default::default()) {
            return Ok(ret);
        }
        if let Some(ret) = write_sync(&self.queue_sync, count, move |i: i32| i) {
            return Ok(ret);
        }

        Err(binder::Status::new_service_specific_error_str(107, Some("no fmq set up")))
    }

    fn getFmqUnsyncWrite(
        &self,
        _: bool,
        _: bool,
        _: &mut MQDescriptor<i32, UnsynchronizedWrite>,
    ) -> BinderResult<bool> {
        // The Rust interface to FMQ does not support `UnsynchronizedWrite`.
        Ok(false)
    }

    /**
     * This method requests the service to trigger a blocking read.
     *
     * @param count Number of messages to read.
     *
     */
    fn requestBlockingRead(&self, count: i32) -> BinderResult<()> {
        self.requestBlockingReadDefaultEventFlagBits(count)
    }

    /**
     * This method requests the service to trigger a blocking read using
     * default Event Flag notification bits defined by the MessageQueue class.
     *
     * @param count Number of messages to read.
     *
     */
    fn requestBlockingReadDefaultEventFlagBits(&self, count: i32) -> BinderResult<()> {
        let mut queue_guard = self.queue_sync.lock().unwrap();
        let Some(ref mut mq) = *queue_guard else {
            return Err(binder::Status::new_service_specific_error_str(107, Some("no fmq set up")));
        };

        let mut buf = vec![0; count as usize];
        let result = mq.read_blocking(&mut buf[..], Some(std::time::Duration::from_secs(5)));

        if !result {
            return Err(binder::Status::new_service_specific_error_str(
                5,
                Some("blocking read failed"),
            ));
        }
        Ok(())
    }

    /**
     * This method requests the service to repeatedly trigger blocking reads.
     *
     * @param count Number of messages to read in a single blocking read.
     * @param numIter Number of blocking reads to trigger.
     *
     */
    fn requestBlockingReadRepeat(&self, count: i32, repeats: i32) -> BinderResult<()> {
        for _ in 0..repeats {
            self.requestBlockingRead(count)?;
        }
        Ok(())
    }

    fn requestReadFmqUnsync(&self, _: i32) -> BinderResult<bool> {
        // The Rust interface to FMQ does not support `UnsynchronizedWrite`.
        Ok(false)
    }
    fn requestWriteFmqUnsync(&self, _: i32) -> BinderResult<bool> {
        // The Rust interface to FMQ does not support `UnsynchronizedWrite`.
        Ok(false)
    }
}
